app_web/src/main.js

```diff
@@ -8,148 +7,380 @@ import { mergeMap, map, share, catchError } from 'rxjs/operators';
+    if (animationPlay) {
+      state.animationTimer.unpause();
+    } else {
+      state.animationTimer.pause();
+    }
+  });
+
+  uiModel.activeAnimations.subscribe(
+    (animations) => (state.animationIndices = animations),
+  );
+  listenForRedraw(uiModel.activeAnimations);
+
+  uiModel.hdr.subscribe((hdrFile) => {
+    resourceLoader.loadEnvironment(hdrFile).then((environment) => {
+      state.environment = environment;
+      // We need to wait until the environment is loaded to redraw
+      redraw = true;
     });
-
-    // Only redraw glTF view upon user inputs, or when an animation is playing.
-    let redraw = false;
-    const listenForRedraw = stream => stream.subscribe(() => redraw = true);
-    
-    uiModel.scene.subscribe(scene => state.sceneIndex = scene !== -1 ? scene : undefined);
-    listenForRedraw(uiModel.scene);
-
-    uiModel.camera.subscribe(camera => state.cameraIndex = camera !== -1 ? camera : undefined);
-    listenForRedraw(uiModel.camera);
-
-    uiModel.variant.subscribe(variant => state.variant = variant);
-    listenForRedraw(uiModel.variant);
-
-    uiModel.tonemap.subscribe(tonemap => state.renderingParameters.toneMap = tonemap);
-    listenForRedraw(uiModel.tonemap);
-
-    uiModel.debugchannel.subscribe(debugchannel => state.renderingParameters.debugOutput = debugchannel);
-    listenForRedraw(uiModel.debugchannel);
-
-    uiModel.skinningEnabled.subscribe(skinningEnabled => state.renderingParameters.skinning = skinningEnabled);
-    listenForRedraw(uiModel.skinningEnabled);
+  });
+
+  uiModel.attachGltfLoaded(gltfLoaded);
+  uiModel.updateStatistics(statisticsUpdateObservable);
+  const sceneChangedStateObservable = uiModel.scene.pipe(map(() => state));
+  uiModel.attachCameraChangeObservable(sceneChangedStateObservable);
+
+  uiModel.orbit.subscribe((orbit) => {
+    if (state.cameraIndex === undefined) {
+      state.userCamera.orbit(orbit.deltaPhi, orbit.deltaTheta);
+    }
+  });
+  listenForRedraw(uiModel.orbit);
+
+  uiModel.pan.subscribe((pan) => {
+    if (state.cameraIndex === undefined) {
+      state.userCamera.pan(pan.deltaX, -pan.deltaY);
+    }
+  });
+  listenForRedraw(uiModel.pan);
+
+  uiModel.zoom.subscribe((zoom) => {
+    if (state.cameraIndex === undefined) {
+      state.userCamera.zoomBy(zoom.deltaZoom);
+    }
+  });
+  listenForRedraw(uiModel.zoom);
+
+  // configure the animation loop
+  const past = {};
+  const update = () => {
+    const devicePixelRatio = window.devicePixelRatio || 1;
+
+    // set the size of the drawingBuffer based on the size it's displayed.
+    canvas.width = Math.floor(canvas.clientWidth * devicePixelRatio);
+    canvas.height = Math.floor(canvas.clientHeight * devicePixelRatio);
+    redraw |= !state.animationTimer.paused && state.animationIndices.length > 0;
+    redraw |= past.width != canvas.width || past.height != canvas.height;
+    past.width = canvas.width;
+    past.height = canvas.height;
```

Comment:
There is an `update `function defined here which in turn has an `update `function inside of it, as well as a repetition of the whole `listenForRedraw `functionality.
This seems overly complicated if I am not mistaken, is this intended?

Also the `update `function and the `past `variable that it references are named ambiguously and must be renamed to avoid access of the wrong symbol (if a duplication of these is even necessary).